### PR TITLE
Immutables library

### PR DESCRIPTION
**Pull request by Mufa0**
Please read through the [writing guide](https://reflectoring.io/contribute/writing-guide/) before submitting a pull request and go through this checklist (at least for your first article).

**Feel free to raise the pull request even if the checklist hasn't been worked through, yet**, to take advantage of the online preview that is generated for each pull request (see "deploy/netlify" under "Show all checks" in your pull request). But please only flag it to be reviewed by your editor after going through the checklist.

## Review
- [x] you have reviewed the text a day after finishing it (you'd be surprised how many errors you can avoid by just getting some distance to what you've written)

## Content
- [x] the article explains *why* and not only *how*
- [x] (half-)sentences that express a meaningful idea that is understandable by itself own are **bold** to allow quick scanning of the text
- [x] the text has been checked with [Grammarly](https://grammarly.com) (the free tier is enough)

## Text
- [x] the text is conversational ("we" instead of "you", easy language)
- [x] there are no "walls of text" (short sentences, short paragraphs with a single idea)
- [x] the text is inclusive ("the developers" and "they" instead of "the developer" and "he/she")
- [x] the text uses active voice instead of passive voice
- [x] names are spelled consistently throughout the article (correct uppercase / lowercase)
- [x] file names and variables are highlighted as `code`

## Structure
- [x] the article has a short intro before the first headline
- [x] by reading the table of contents (TOC) alone, I get a sense of the "story" of the article (check the TOC in the preview after raising the PR or by starting the blog locally on your machine)
- [x] if using sub-headlines, there's at least two sub-headlines below a main-headline
- [x] headlines are consistent throughout the article (same level of abstraction, same style - if one headline starts with a verb in imperative form, the other headlines on the same level should probably, too)

## Code Examples
- [x] code examples are introduced with a ":" and explained below the code example
- [x] code examples are formatted to avoid horizontal scrolling (manual line breaks and 2-spaces indentation)

### CODE REVIEW

File: _posts/2022-01-14-immutables-library.md [hunk 1]
```diff
@@ -0,0 +1,793 @@
...
+        allParameters = true,
+        visibility = Value.Style.ImplementationVisibility.PUBLIC
+
+)
+public @interface CustomStyle {}
+```
+After defining `@Target` and `@Retention` annotation, we come to the `@Value.Style` annotation. The first value defined that we want to use the `new` constructor. The next thing that we define is that we want to use the strictBuilder and that all attributes should be annotated with the `@Value.Parameter` annotation. The last style defined is that the implementation visibility will be public.
+
+For more information about style possibilities, please refer to the [official documentation.](https://immutables.github.io/style.html)
+
```
Comment (thombergs) [inline _posts/2022-01-14-immutables-library.md]: Add a code example that shows how to use this meta annotation.

File: _posts/2022-01-14-immutables-library.md [hunk 2]
```diff
@@ -0,0 +1,793 @@
...
+# Conclusion
+We saw how the immutable library helps us build immutable, thread-safe, and null-safe domain objects. It helps us build clean and readable POJO classes. 
+
+Since it is a powerful tool we need to be careful how to use it. We can easily stray down the wrong path and overuse its features.
```
Comment (thombergs) [inline _posts/2022-01-14-immutables-library.md]: Why is overusing its features bad? Add a sentence explaining that. Or, even better, add a whole section explaining what can go wrong.

File: _posts/2022-01-14-immutables-library.md [hunk 3]
```diff
@@ -0,0 +1,793 @@
...
+
+The last thing that I want to point out is the `@Value.Style` annotation. The `@Value.Immutable` annotation tells what will be generated, while the `@Value.Style` tells how will be generated. This annotation can be a slippery slope, and we need to be careful and go outside of the default setting only when we are certain that we need to.
+
+For deeper reading on the immutable library please refer to the [official page.](https://immutables.github.io/)
```
Comment (thombergs) [inline _posts/2022-01-14-immutables-library.md]: Add a link to your code examples again as a last sentence so readers can look up the example code there.

**Review by thombergs — CHANGES_REQUESTED** [review]
This is a very well-structured article! I made some edits:

- removed some self-explanatory code and sentences
- renamed some sections to make them more consistent
- renamed "immutable library" to "Immutables library" everywhere (careful with names)

For a great article, what's missing is some explanations as to WHY we should do things one way or another. I left some comments on the PR regarding this. Please have a look through the comments and decide for each one whether you want to action it or not ....

File: _posts/2022-01-14-immutables-library.md [hunk 4]
```diff
@@ -0,0 +1,793 @@
...
+    abstract String getContent();
+
+    abstract long getUserId();
+}
+```
+The first thing that we notice is the special annotation. 
+
+The `@Value.Immutable` annotation says to the annotation processor that it should generate an implementation for this class.
+
+It is important to mention that we can place the `@Value.Immutable` annotation on the class, interface or annotation type.
```
[the remark this answers is not in the document]
Comment (Mufa0) [reply _posts/2022-01-14-immutables-library.md]: When adding Immutable to the annotation we the immutable implementation of that interface will be generated, but I don't know the real usage. 
I see two options:
1. I remove this part of adding @Value.Immutable to the annotation
2. I leave this part but then do you know the real usage of using the annotation like this:
        CustomAnnotation customAnnotation =
                ImmutableCustomAnnotation.builder()
                        .values("a")
                        .build();
        customAnnotation.values();

File: _posts/2022-01-14-immutables-library.md [hunk 5]
```diff
@@ -0,0 +1,793 @@
...
+    abstract String getTitle();
+
+    abstract String getContent();
+
+    abstract long getUserId();
+}
+```
+The first thing that we notice is the special annotation. 
+
+The `@Value.Immutable` annotation says to the annotation processor that it should generate an implementation for this class.
```
[the remark this answers is not in the document]
Comment (Mufa0) [reply _posts/2022-01-14-immutables-library.md]: I didn't have to enable an annotation processor in the IntelliJ and I just checked if it was checked from before. It was not.

File: _posts/2022-01-14-immutables-library.md [hunk 6]
```diff
@@ -0,0 +1,793 @@
...
+Exception in thread "main" java.lang.UnsupportedOperationException
+at java.base/java.util.Collections$UnmodifiableCollection.add(Collections.java:1060)
+at com.reflectoring.io.immutables.collections.CollectionsService.main(CollectionsService.java:45)
+```
+After adding the new article on the already built user, we get the UnsupportedOperationException. After building, the list is immutable, and we cannot add anything new to it. If we want to expand this list, we need to create a new user.
+
+# Styles
+The `@Value.Style` is the annotation with which we control what will the annotation process generate. In the [prevous example](#using-the-new-method), we used the `@Value.Style` annotation to generate the standard constructor format. 
+
+We can use the annotation on several levels:
```
[the remark this answers is not in the document]
Comment (Mufa0) [reply _posts/2022-01-14-immutables-library.md]: Did you mean that I should list all available styles?

**Pull request merged**